### PR TITLE
Feat: Add ra-auth0 package + demo

### PR DESCRIPTION
**Pull request by arimet**

### CODE REVIEW

**Comment by fzaninotto** [comment]
Could you rename the repo & package `ra-auth-auth0`?

**Comment by djhi** [comment]
Why don't we include it in the main repo?

**Comment by fzaninotto** [comment]
> Why don't we include it in the main repo?

@djhi to have more freedom over the release schedule. Lerna forces us to release almost all packages as soon as we update `ra-core` - I'm not fond of such "shadow" releases.

**Comment by djhi** [comment]
Isn't it a configuration issue ?

**Comment by fzaninotto** [comment]
Right, but we want to keep a concurrent release for the main packages to ease installation (ra-core, ra-ui-materialui, react-admin) and decorrelate the other packages. Lerna is all-or-nothing. So using another repo is a good solution here IMHO.

File: README.md [hunk 1]
```diff
@@ -1 +1,82 @@
...
+
+For react-admin, you need to setup environment variables. You can do this by creating a `.env` file in the root of the project. The following variables are required:
+
+```JS
+VITE_AUTH0_DOMAIN=your-domain.auth0.com
+VITE_AUTH0_CLIENT_ID=your-client-id
+VITE_AUTH0_AUDIENCE=your-audience // optional but recommended for aving non opaque tokens
+VITE_LOGIN_REDIRECT_URL = 'http://127.0.0.1:8081'
+VITE_LOGOUT_REDIRECT_URL = 'http://127.0.0.1:8081'
+VITE_API_URL = 'http://127.0.0.1:3000'
```
Comment (slax57) [inline README.md]: These should be in the actual `.env.example` file IMO

File: README.md [hunk 2]
```diff
@@ -1 +1,82 @@
...
+```
+
+The `authProvider` exported by this package will look for a claim that has a name with the term `role` in it and return its value.
+
+For react-admin, you need to setup environment variables. You can do this by creating a `.env` file in the root of the project. The following variables are required:
+
+```JS
+VITE_AUTH0_DOMAIN=your-domain.auth0.com
+VITE_AUTH0_CLIENT_ID=your-client-id
+VITE_AUTH0_AUDIENCE=your-audience // optional but recommended for aving non opaque tokens
```
Comment (slax57) [inline README.md]:
```suggestion
VITE_AUTH0_AUDIENCE=your-audience // optional but recommended for having non opaque tokens
```

File: README.md [hunk 3]
```diff
@@ -1 +1,82 @@
...
+VITE_LOGIN_REDIRECT_URL = 'http://127.0.0.1:8081'
+VITE_LOGOUT_REDIRECT_URL = 'http://127.0.0.1:8081'
+VITE_API_URL = 'http://127.0.0.1:3000'
+```
+
+1. Run `make install run` to install the dependencies and start the Demo App
```
Comment (slax57) [inline README.md]: There is no `run` target in the `makefile`

File: README.md [hunk 4]
```diff
@@ -1 +1,82 @@
...
+Now that all is configured and running, you can browse to http://localhost:8081/ to access the React Admin App.
+
+-   Signing in with `user@marmelab.com` will only grant the `user` role permissions
+-   Signing in with `admin@marmelab.com` will grant full `admin` role permissions, allowing for instance to see the 'Users' resource in the main menu
+
+Feel free to play around with this demo, along with the Auth0 config, to understand better how it works!
+
+### JSON Server API
+
+The demo app uses a [JSON Server]() to provide a REST API. You can find the configuration in the `demo-fake-api` folder.
```
Comment (slax57) [inline README.md]: missing link

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+version: '3'
+
+volumes:
+    postgres_data:
+        driver: local
+
+services:
+    # Create service demo-fake-api who user node 18 on port 3000 and code is in packages/demo-fake-api folder. Without postgress
```
Comment (slax57) [inline docker-compose.yml]:
```suggestion
    # Create service demo-fake-api who uses node 18 on port 3000 and code is in packages/demo-fake-api folder. Without postgres
```

And I don't understand the 'Without postgres' part

File: packages/ra-auth-auth0/Readme.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+npm install --save ra-auth-auth0
+```
+
+## Example usage
+
+```jsx
+// in src/App.tsx
+import React, { useEffect, useRef, useState } from 'react';
+import { Admin, Resource } from 'react-admin';
+import { Auth0AuthProvider, httpClient } from 'ra-auth-auth0';
```
Comment (slax57) [inline packages/ra-auth-auth0/Readme.md]: `httpClient` is not used in this example

File: packages/ra-auth-auth0/Readme.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+## `Auth0AuthProvider` Parameters
+
+-   `loginRedirectUri` - _optional_ - URI used to override the redirect URI after successful login. Defaults to react-admin `/auth-callback` route.
+-   `logoutRedirectUri` - _optional_ - URI used to override the redirect URI after successful logout. Defaults to the current url.
+-   `redirectOnCheckAuth` - _optional_ - If set to `true` (the default), redirect users to the Auth0 login page inside the [`checkAuth`](https://marmelab.com/react-admin/AuthProviderWriting.html#checkauth) method. Set it to `false` if you want users to go through a [custom login page](https://marmelab.com/react-admin/Authentication.html#customizing-the-login-component) first.
+
+## Passing The Auth0 Token To Your Backend
+
+If you want to pass the Auth0 authentication token to your backend, you can use the [`httpClient`](https://marmelab.com/react-admin/DataProviders.html#adding-custom-headers) exported by package. For instance, here's how to use it with the [`ra-data-json-server` dataProvider](https://github.com/marmelab/react-admin/tree/master/packages/ra-data-json-server):
```
Comment (slax57) [inline packages/ra-auth-auth0/Readme.md]:
```suggestion
If you want to pass the Auth0 authentication token to your backend, you can use the [`httpClient`](https://marmelab.com/react-admin/DataProviders.html#adding-custom-headers) exported by this package. For instance, here's how to use it with the [`ra-data-json-server` dataProvider](https://github.com/marmelab/react-admin/tree/master/packages/ra-data-json-server):
```

File: packages/ra-auth-auth0/src/authProvider.ts [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+ *          });
+ *
+ *          const httpClientAuth0 = await httpClient(clientAuth0);
+ *          dataProvider.current = jsonServerProvider(
+ *              import.meta.env.VITE_API_URL,
+ *              httpClientAuth0
+ *          );
+ *
+ *          setAuth0(clientAuth0);
+ *       };
```
Comment (slax57) [inline packages/ra-auth-auth0/src/authProvider.ts]: Code example does not seem up to date with what is inside the Readme and the Demo

File: packages/ra-auth-auth0/src/authProvider.ts [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+ *               </>
+ *           )}
+ *       </Admin>
+ *    );
+ * };
+ * export default App;
+ *
+ * ```
+ *
+ * @param client the Auth0 client *
```
Comment (slax57) [inline packages/ra-auth-auth0/src/authProvider.ts]:
```suggestion
 * @param client the Auth0 client
```

Besides other possible params are missing from the JSDoc

**Review by slax57 — CHANGES_REQUESTED** [review]

File: packages/ra-auth-auth0/Readme.md (hunk 6 above)
Comment (djhi) [reply packages/ra-auth-auth0/Readme.md]:
```suggestion
import { Auth0AuthProvider } from 'ra-auth-auth0';
```

File: README.md (hunk 3 above)
Comment (djhi) [reply README.md]:
```suggestion
1. Run `make install start` to install the dependencies and start the Demo App
```

File: README.md (hunk 1 above)
Comment (djhi) [reply README.md]: They are?

File: docker-compose.yml (hunk 5 above)
Comment (slax57) [reply docker-compose.yml]: > And I don't understand the 'Without postgres' part

What about this?

Comment (djhi) [reply docker-compose.yml]: me neither, I haven't read all files added by Anthony

File: packages/ra-auth-auth0/src/authProvider.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+import { Auth0Client } from '@auth0/auth0-spa-js';
+
+/**
+ * An authProvider which handles authentication via the Auth0 instance.
+ *
+ * @example
+ * ```tsx
+ * import React from 'react';
+ * import { Admin, Resource } from 'react-admin';
+ * import { Route } from 'react-router-dom';
```
Comment (slax57) [inline packages/ra-auth-auth0/src/authProvider.ts]: unused import

File: packages/ra-auth-auth0/src/authProvider.ts [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+ * import { Auth0AuthProvider, httpClient } from 'ra-auth-auth0';
```
Comment (slax57) [inline packages/ra-auth-auth0/src/authProvider.ts]: `httpClient` is unused

File: packages/ra-auth-auth0/src/authProvider.ts [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+ * import { Auth0Client } from '@auth0/auth0-spa-js';
+ * import dataProvider from './dataProvider';
+ * import posts from './posts';
+ *
+ * const clientAuth0 = new Auth0Client({
+ *    domain: 'your-domain.auth0.com',
```
Comment (slax57) [inline packages/ra-auth-auth0/src/authProvider.ts:17]: why hard-code some values here, but still use env vars in `loginRedirectUri: import.meta.env.VITE_LOGIN_REDIRECT_URL` a few lines below?

**Review by slax57 — CHANGES_REQUESTED** [review]
Almost there :slightly_smiling_face:

File: README.md [hunk 13]
```diff
@@ -1 +1,82 @@
...
+1. Select your Auth0 Application (or create a new one)
+1. Add `admin@marmelab.com` user to your Auth0 Application. For this, you need to go to the `User Management` section and create a new user.
+1. Add `admin` to the `Roles` section of your Auth0 Application. Assign admin@marmelab.com to the `admin` role.
+1. Do the same for `user@marmelab.com` and the `user` role.
+1. Go to the `Applications` section of your Auth0 Application and select `Settings`
+1. Create a `single page` application with the following settings:
+    - Allowed Callback URLs: `http://localhost:8081`
+    - Allowed Logout URLs: `http://localhost:8081`
+    - Allowed Web Origins: `http://localhost:8081`
+    - Allowed Origins (CORS): `http://localhost:8081`
```
Comment (slax57) [inline README.md]: Addresses are `http://127.0.0.1:8081` if your run from Docker.
Also the callback URL is wrong.

```suggestion
    - Allowed Callback URLs: `http://127.0.0.1:8081/auth-callback`
    - Allowed Logout URLs: `http://127.0.0.1:8081`
    - Allowed Web Origins: `http://127.0.0.1:8081`
    - Allowed Origins (CORS): `http://127.0.0.1:8081`
```

File: .env.example [hunk 14]
```diff
@@ -0,0 +1,6 @@
+VITE_AUTH0_DOMAIN="your-domain.auth0.com"
+VITE_AUTH0_CLIENT_ID="your-client-id"
+VITE_AUTH0_AUDIENCE="https://your-domain.auth0.com/api/v2/" # optional but recommended for having non opaque tokens
+VITE_LOGIN_REDIRECT_URL="http://127.0.0.1:8081/auth-callback"
+VITE_LOGOUT_REDIRECT_URL="http://127.0.0.1:8081"
+VITE_API_URL="http://127.0.0.1:3000"
```
Comment (slax57) [inline .env.example]: `VITE_API_URL` is never used

File: packages/ra-auth-auth0/src/authProvider.ts [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    async handleCallback() {
+        const query = window.location.search;
+        if (query.includes('code=') && query.includes('state=')) {
+            try {
+                await client.handleRedirectCallback();
+                return;
+            } catch (error) {
+                console.log('error', error);
+                throw error;
+            }
```
Comment (slax57) [inline packages/ra-auth-auth0/src/authProvider.ts:145]: Error handling here is broken
It triggers an infinite loop unless we set `logoutOnFailure=false`.

Temporary (dirty) workaround:

```suggestion
            } catch (error) {
                console.log('error', error);
                // eslint-disable-next-line no-throw-literal
                throw { message: error.message, logoutOnFailure: false };
            }
```

It probably comes from one of the `queryClient.clear();` in [useLogout](https://github.com/marmelab/react-admin/blob/next/packages/ra-core/src/auth/useLogout.ts#L69).

**Review by slax57 — CHANGES_REQUESTED** [review]
Two more things:
- We should probably keep a TODO somewhere to indicate that this package needs the upcoming react-admin v4.7 to run
- `yarn.lock` is not up to date (it gets modified after running `make install`)

File: .env.example (hunk 14 above)
Comment (slax57) [inline .env.example]: `VITE_API_URL` is still referenced here

File: README.md [hunk 16]
```diff
@@ -1 +1,81 @@
...
+
+1. Clone this project
+
+We need to add some minimal configuration to our Auth0 instance to use it. This need to be done from the Auth0 Admin Console.
+
+1. Go to the [Auth0 Dashboard](https://manage.auth0.com/dashboard)
+1. Select your Auth0 Application (or create a new one)
+1. Add `admin@marmelab.com` user to your Auth0 Application. For this, you need to go to the `User Management` section and create a new user.
+1. Add `admin` to the `Roles` section of your Auth0 Application. Assign admin@marmelab.com to the `admin` role.
+1. Do the same for `user@marmelab.com` and the `user` role.
```
Comment (slax57) [inline README.md]: We should maybe mention other emails than `admin@marmelab.com` and `user@marmelab.com` otherwise we will be spammed each time a dev tries to set up Auth0 with this demo :sweat_smile:

File: docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+            - AUTH0_DOMAIN=${VITE_AUTH0_DOMAIN}
+            - AUTH0_AUDIENCE=${VITE_AUTH0_AUDIENCE}
+    demo-react-admin:
+        image: node:18
+        working_dir: /app/packages/demo-react-admin
+        command: yarn start
+        volumes:
+            - '.:/app'
+        environment:
+            - VITE_API_URL=${VITE_API_URL}
```
Comment (slax57) [inline docker-compose.yml]: `VITE_API_URL` is still referenced here

**Review by slax57 — CHANGES_REQUESTED** [review]

File: packages/demo-react-admin/package.json [hunk 18]
```diff
@@ -19,6 +19,7 @@
         "lodash": "~4.17.5",
         "prop-types": "^15.7.2",
         "proxy-polyfill": "^0.3.0",
+        "ra-auth-auth0": "*",
```
Comment (slax57) [inline packages/demo-react-admin/package.json:22]: FMI, why did you need this change?

Comment (djhi) [reply packages/demo-react-admin/package.json:22]: This ensure yarn workspace & vite correctly symlink the dependency in dev mode

**Review by slax57 — APPROVED** [review]
